### PR TITLE
Distributed Autograd: Allow multiple backward passes to accumulate gradients.

### PR DESCRIPTION
**Pull request by pritamdamania87**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#32506 Distributed Autograd: Allow multiple backward passes to accumulate gradients.**

In this PR, we've introduced a `retain_graph` parameter to distributed
autograd similar to `torch.autograd.backward`.

In terms of design, this parameter is sent over RPC to all nodes and is used to
create the GraphTask on the local nodes. This enables us to run
`dist_autograd.backward()` multiple times in the same context.

The use case currently for this is to benchmark only the backward pass for
distributed autograd. We'd like to measure the QPS for the backward pass and as
a result, running a single forward pass and multiple backward passes in a loop
is one way to benchmark backward pass performance.

Differential Revision: [D19521288](https://our.internmc.facebook.com/intern/diff/D19521288/)

### CODE REVIEW

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit 6fdf31ac:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=6fdf31ac8e508a09d26453b58b94de14d71afdbb).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4318581) caffe2_onnx_py3_6_clang7_ubuntu16_04_build (1/1)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=89525565) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=89525565))
<details>
<summary>
<code>Jan 22 20:57:07 /var/lib/jenkins/workspace/test/cpp/dist_autograd/test_dist_autograd.cpp:93:56: error: too few arguments to function call, expected 3, have 2</code>
</summary>

```
Jan 22 20:57:06                                            ^ 
Jan 22 20:57:06 1 warning generated. 
Jan 22 20:57:06 [ 94%] Building CXX object test_api/CMakeFiles/test_api.dir/serialize.cpp.o 
Jan 22 20:57:06 [ 94%] Building CXX object test_jit/CMakeFiles/test_jit.dir/test_graph_executor.cpp.o 
Jan 22 20:57:07 /var/lib/jenkins/workspace/test/cpp/dist_autograd/test_dist_autograd.cpp:70:21: error: too few arguments to function call, expected 2, have 1 
Jan 22 20:57:07   engine.execute({t}); 
Jan 22 20:57:07   ~~~~~~~~~~~~~~    ^ 
Jan 22 20:57:07 /var/lib/jenkins/workspace/torch/csrc/distributed/autograd/engine/dist_engine.h:36:3: note: 'execute' declared here 
Jan 22 20:57:07   void execute(const torch::autograd::variable_list& roots, bool retainGraph); 
Jan 22 20:57:07   ^ 
Jan 22 20:57:07 /var/lib/jenkins/workspace/test/cpp/dist_autograd/test_dist_autograd.cpp:93:56: error: too few arguments to function call, expected 3, have 2 
Jan 22 20:57:07   engine.executeSendFunctionAsync(context, sendFunction)->wait(); 
Jan 22 20:57:07   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~                      ^ 
Jan 22 20:57:07 /var/lib/jenkins/workspace/torch/csrc/distributed/autograd/engine/dist_engine.h:44:3: note: 'executeSendFunctionAsync' declared here 
Jan 22 20:57:07   std::shared_ptr<rpc::FutureMessage> executeSendFunctionAsync( 
Jan 22 20:57:07   ^ 
Jan 22 20:57:07 2 errors generated. 
Jan 22 20:57:07 dist_autograd/CMakeFiles/test_dist_autograd.dir/build.make:86: recipe for target 'dist_autograd/CMakeFiles/test_dist_autograd.dir/test_dist_autograd.cpp.o' failed 
Jan 22 20:57:07 make[2]: *** [dist_autograd/CMakeFiles/test_dist_autograd.dir/test_dist_autograd.cpp.o] Error 1 
Jan 22 20:57:07 CMakeFiles/Makefile2:11541: recipe for target 'dist_autograd/CMakeFiles/test_dist_autograd.dir/all' failed 
Jan 22 20:57:07 make[1]: *** [dist_autograd/CMakeFiles/test_dist_autograd.dir/all] Error 2 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

**Comment by albanD** [comment]
The CI failure look relevant right?

Also high level question (using pseudocode below) before looking into the code in detail:
```python
# loss contains a graph
# assume that .backward() is asynchronous and we need to call .wait()

# This is fine because the first one is done before the second starts
loss.backward(retain_graph=True).wait()
loss.backward().wait()

# What happens here if the second backward gets ahead of the first one and
# deletes the graph before the first one is done with it?
# This will lead to flaky code that might or might not fail.
# What is the strategy to fix this?
fut1 = loss.backward(retain_graph=True)
fut2 = loss.backward()
fut1.wait()
fut2.wait()

```

**Comment by pritamdamania87** [comment]
> The CI failure look relevant right?
> 
> Also high level question (using pseudocode below) before looking into the code in detail:
> 
> ```python
> # loss contains a graph
> # assume that .backward() is asynchronous and we need to call .wait()
> 
> # This is fine because the first one is done before the second starts
> loss.backward(retain_graph=True).wait()
> loss.backward().wait()
> 
> # What happens here if the second backward gets ahead of the first one and
> # deletes the graph before the first one is done with it?
> # This will lead to flaky code that might or might not fail.
> # What is the strategy to fix this?
> fut1 = loss.backward(retain_graph=True)
> fut2 = loss.backward()
> fut1.wait()
> fut2.wait()
> ```

Will look into the CI failures. Regarding the async backward, although we have provided an async option in autograd/engine.cpp, we don't have any async API for backward/autograd exposed the user. Right now calling backward() from python is blocking for both local autograd and distributed autograd. Also, note Engine::execute_with_graph_task (used by distributed autograd) is async whereas Engine::execute() is still a blocking call.

**Comment by albanD** [comment]
I know this is not currently possible to do from python (above was "pseudocode")
But using distributed it might be possible to do something similar using the provided api?

**Comment by pritamdamania87** [comment]
> But using distributed it might be possible to do something similar using the provided api?
It is possible for PyTorch developers to do something like this using the API (not sure how end users could do something like this) to maybe build something on top of this, although I don't think we should support this. In general I feel this API is mostly an internal API that we tightly control and can ensure we don't end up using the API in this way to build on top of this. 

Also, note that this is only possible if you call Engine::execute_with_graph_task directly. Currently, distributed autograd doesn't provide any async API either in C++/python (it just used the async API of the local autograd engine internally). We can add a comment to `Engine::execute_with_graph_task` mentioning that we should use the API carefully and currently only the distributed autograd use case is what we support.

**Comment by pritamdamania87** [comment]
@albanD I've fixed the test failures, could you take another look? Thanks!

File: torch/csrc/autograd/engine.cpp [hunk 1]
```diff
@@ -154,8 +154,14 @@ struct ReadyQueue {
...
   std::shared_ptr<GraphTask> graph_task = base_.lock();
-  TORCH_INTERNAL_ASSERT(graph_task, "GraphTask is no longer valid!")
-  return graph_task->reentrant_depth_;
+  if (graph_task) {
+    return graph_task->reentrant_depth_;
+  } else {
+    // The graph task is no longer valid indicating an error. As a result, we
+    // try to move this to the front of the queue to ensure the autograd
+    // engine threads pick up this error soon.
+    return std::numeric_limits<int>::max();
```
Comment (albanD) [inline torch/csrc/autograd/engine.cpp]: nit: Maybe we want to introduce some different levels of "max" here. I think this soon be higher priority than the AccumulateGrad [here](https://github.com/pytorch/pytorch/blob/6e38a709397ba2299e99011618f78dd5ae4c048a/torch/csrc/autograd/functions/accumulate_grad.cpp#L20) that use `UINT64_MAX`.

File: torch/csrc/distributed/autograd/context/context.cpp [hunk 2]
```diff
@@ -74,7 +74,12 @@ void DistAutogradContext::accumulateGrad(
     it->value().add_(grad);
   } else {
     // First grad for this variable.
-    accumulatedGrads_.insert(variable, grad);
+    if (grad.is_sparse()) {
+      accumulatedGrads_.insert(variable, grad.clone());
```
Comment (albanD) [inline torch/csrc/distributed/autograd/context/context.cpp]: nit: Why does the clone below has a memory format but not this one?

File: torch/csrc/distributed/autograd/functions/recvrpc_backward.cpp [hunk 3]
```diff
@@ -30,21 +30,23 @@ variable_list RecvRpcBackward::apply(variable_list&& grads) {
     }
   }
 
+  auto sharedContext = autogradContext_.lock();
+  TORCH_CHECK(sharedContext, "Autograd context no longer valid");
```
Comment (albanD) [inline torch/csrc/distributed/autograd/functions/recvrpc_backward.cpp]: nit: TORCH_INTERNAL_ASSERT ?

**Review by albanD — APPROVED** [review]
Looks good. Small details only.

File: torch/csrc/autograd/engine.cpp (hunk 1 above)
Comment (pritamdamania87) [reply torch/csrc/autograd/engine.cpp]: Doesn't AccumulateGrad use sequence_nr, but this is actually using reentrant_depth? Looking at CompareNodeTaskTime, it looks like we look at sequence_nr only if reentrant_depth is same. So this will be higher priority than AccumulateGrad.

I think we can add more fine grained levels when we have another case similar to this.

File: torch/csrc/distributed/autograd/context/context.cpp (hunk 2 above)
Comment (pritamdamania87) [reply torch/csrc/distributed/autograd/context/context.cpp]: This pretty much follows what we do for AccumulateGrad: https://github.com/pytorch/pytorch/blob/6e38a709397ba2299e99011618f78dd5ae4c048a/torch/csrc/autograd/functions/accumulate_grad.cpp#L57. We use contiguous below since the grads are not sparse, but here we don't specify any memory format which defaults to preserving the memory format of the original tensor.

File: torch/csrc/distributed/autograd/functions/recvrpc_backward.cpp (hunk 3 above)
Comment (pritamdamania87) [reply torch/csrc/distributed/autograd/functions/recvrpc_backward.cpp]: This isn't really a bug if this check fails. It is possible that this check fails because distributed autograd failed and another thread cleaned up the autograd context. In this case, its fine for this line to throw an exception to reporting an error upstream.

File: torch/csrc/autograd/engine.cpp (hunk 1 above)
Comment (albanD) [reply torch/csrc/autograd/engine.cpp]: Ho right. Sounds good then.
This is a bit hacky but that will work.

File: torch/csrc/distributed/autograd/context/context.cpp (hunk 2 above)
Comment (albanD) [reply torch/csrc/distributed/autograd/context/context.cpp]: But I think AccumulateGrad does that mostly for historical reasons but it is fine to just follow the same behavior.

File: torch/csrc/distributed/autograd/functions/recvrpc_backward.cpp (hunk 3 above)
Comment (albanD) [reply torch/csrc/distributed/autograd/functions/recvrpc_backward.cpp]: >  its fine for this line to throw an exception to reporting an error upstream

Not sure to follow this sentence?

If this is a user facing error, it should be a bit more explicit no?
The internal assert will tell the user to report a bug to pytorch. But a regular one will just say that. Is the idea that the user should look at the other processes for the real error?

Comment (pritamdamania87) [reply torch/csrc/distributed/autograd/functions/recvrpc_backward.cpp]: Yes, the idea is that the user should look at other processes for the real error. You have a good point, I'll make this error more explicit.
